File: core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/user/RegisterUser.java
package org.mifos.mobilewallet.core.domain.usecase.user;

import org.mifos.mobilewallet.core.base.UseCase;
import org.mifos.mobilewallet.core.data.fineract.entity.register.RegisterPayload;
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository;
import org.mifos.mobilewallet.core.utils.Constants;

import javax.inject.Inject;

import okhttp3.ResponseBody;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by naman on 17/8/17.
 */

public class RegisterUser extends UseCase<RegisterUser.RequestValues,
        RegisterUser.ResponseValue> {

    private final FineractRepository apiRepository;

    @Inject
    public RegisterUser(FineractRepository apiRepository) {
        this.apiRepository = apiRepository;
    }


    @Override
    protected void executeUseCase(RequestValues requestValues) {

        apiRepository.registerUser(requestValues.registerPayload)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<ResponseBody>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        getUseCaseCallback().onError(Constants.ERROR_REGISTERING_USER);
                    }

                    @Override
                    public void onNext(ResponseBody responseBody) {
                        getUseCaseCallback().onSuccess(new ResponseValue());
                    }
                });

    }


    public static final class RequestValues implements UseCase.RequestValues {

        private final RegisterPayload registerPayload;

        public RequestValues(RegisterPayload registerPayload) {
            this.registerPayload = registerPayload;
        }
    }

    public static final class ResponseValue implements UseCase.ResponseValue {

        public ResponseValue() {

        }

    }
}

File: core/data/src/main/java/org/mifos/mobilewallet/core/domain/usecase/user/RegisterUser.kt
package org.mifos.mobilewallet.core.domain.usecase.user

import com.mifos.mobilewallet.model.entity.register.RegisterPayload
import okhttp3.ResponseBody
import org.mifos.mobilewallet.core.base.UseCase
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository
import org.mifos.mobilewallet.core.util.Constants
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by naman on 17/8/17.
 */
class RegisterUser @Inject constructor(private val apiRepository: FineractRepository) :
    UseCase<RegisterUser.RequestValues, RegisterUser.ResponseValue>() {
    protected override fun executeUseCase(requestValues: RequestValues) {
        apiRepository.registerUser(requestValues.registerPayload)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : Subscriber<ResponseBody?>() {
                override fun onCompleted() {}
                override fun onError(e: Throwable) {
                    useCaseCallback.onError(Constants.ERROR_REGISTERING_USER)
                }

                override fun onNext(t: ResponseBody?) {
                    useCaseCallback.onSuccess(ResponseValue())
                }
            })
    }

    data class RequestValues(val registerPayload: RegisterPayload) : UseCase.RequestValues
    class ResponseValue : UseCase.ResponseValue
}
